File: sdk/go/identity.go
package alpgo

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type AgentIdentity struct {
	DID        string
	AgentID    string
	PublicKey  string
	CreatedAt  string
	Metadata   map[string]any
}

func NewAgentIdentity(did, agentID, publicKey string, createdAt string, metadata map[string]any) *AgentIdentity {
	if createdAt == "" {
		createdAt = time.Now().Format(time.RFC3339)
	}
	if metadata == nil {
		metadata = make(map[string]any)
	}
	return &AgentIdentity{
		DID:        did,
		AgentID:    agentID,
		PublicKey:  publicKey,
		CreatedAt:  createdAt,
		Metadata:   metadata,
	}
}

func (i *AgentIdentity) ToDict() map[string]any {
	return map[string]any{
		"did":        i.DID,
		"agent_id":   i.AgentID,
		"public_key": i.PublicKey,
		"created_at": i.CreatedAt,
		"metadata":   i.Metadata,
	}
}

func AgentIdentityFromDict(d map[string]any) *AgentIdentity {
	did, _ := d["did"].(string)
	agentID, _ := d["agent_id"].(string)
	publicKey, _ := d["public_key"].(string)
	createdAt, _ := d["created_at"].(string)
	metadata, _ := d["metadata"].(map[string]any)
	return NewAgentIdentity(did, agentID, publicKey, createdAt, metadata)
}

type VerifiablePresentation struct {
	DID        string
	AgentID    string
	Claims     map[string]any
	Signature  string
	IssuedAt   string
}

func NewVerifiablePresentation(did, agentID string, claims map[string]any, signature, issuedAt string) *VerifiablePresentation {
	if issuedAt == "" {
		issuedAt = time.Now().Format(time.RFC3339)
	}
	return &VerifiablePresentation{
		DID:       did,
		AgentID:   agentID,
		Claims:    claims,
		Signature: signature,
		IssuedAt:  issuedAt,
	}
}

func (v *VerifiablePresentation) ToDict() map[string]any {
	return map[string]any{
		"did":       v.DID,
		"agent_id":  v.AgentID,
		"claims":    v.Claims,
		"signature": v.Signature,
		"issued_at": v.IssuedAt,
	}
}

func (v *VerifiablePresentation) Verify(publicKey string) bool {
	payload := fmt.Sprintf(`{"did":"%s","agent_id":"%s","claims":%v}`, v.DID, v.AgentID, v.Claims)
	expected := sha256Hex(payload + publicKey)
	return v.Signature == expected
}

type TrustEntry struct {
	AgentID     string
	Scopes      []string
	TrustLevel  string
	RegisteredAt string
}

func NewTrustEntry(agentID string, scopes []string, trustLevel, registeredAt string) *TrustEntry {
	if trustLevel == "" {
		trustLevel = "standard"
	}
	if registeredAt == "" {
		registeredAt = time.Now().Format(time.RFC3339)
	}
	return &TrustEntry{
		AgentID:     agentID,
		Scopes:      scopes,
		TrustLevel:  trustLevel,
		RegisteredAt: registeredAt,
	}
}

type TrustRegistry struct {
	AlpDir  string
	Entries map[string]*TrustEntry
}

func NewTrustRegistry(alpDir string) *TrustRegistry {
	t := &TrustRegistry{
		AlpDir:  alpDir,
		Entries: make(map[string]*TrustEntry),
	}
	t.Load()
	return t
}

func (t *TrustRegistry) identityDir() string {
	return filepath.Join(t.AlpDir, ".identity")
}

func (t *TrustRegistry) trustPath() string {
	return filepath.Join(t.identityDir(), "trust_registry.json")
}

func (t *TrustRegistry) Load() {
	p := t.trustPath()
	data, err := os.ReadFile(p)
	if err != nil {
		return
	}
	// Simple JSON parse not available without external lib; use simple line-based storage in production
	_ = data
}

func (t *TrustRegistry) Save() {
	d := t.identityDir()
	if err := os.MkdirAll(d, 0755); err != nil {
		return
	}
	// In production, serialize to JSON
	_ = d
}

func (t *TrustRegistry) Register(did, agentID string, scopes []string, trustLevel string) *TrustEntry {
	entry := NewTrustEntry(agentID, scopes, trustLevel, "")
	t.Entries[did] = entry
	t.Save()
	return entry
}

func (t *TrustRegistry) Resolve(did string) *TrustEntry {
	return t.Entries[did]
}

func (t *TrustRegistry) Revoke(did string) bool {
	if _, ok := t.Entries[did]; ok {
		delete(t.Entries, did)
		t.Save()
		return true
	}
	return false
}

func (t *TrustRegistry) ListDIDs() []string {
	dids := make([]string, 0, len(t.Entries))
	for did := range t.Entries {
		dids = append(dids, did)
	}
	return dids
}

func (t *TrustRegistry) HasScope(did, requiredScope string) bool {
	entry := t.Entries[did]
	if entry == nil {
		return false
	}
	for _, s := range entry.Scopes {
		if s == requiredScope {
			return true
		}
	}
	return false
}

type IdentityResolver struct {
	TrustRegistry *TrustRegistry
}

func NewIdentityResolver(trustRegistry *TrustRegistry) *IdentityResolver {
	return &IdentityResolver{TrustRegistry: trustRegistry}
}

func (r *IdentityResolver) VerifyPresentation(presentation *VerifiablePresentation, publicKey string) map[string]any {
	if !presentation.Verify(publicKey) {
		return map[string]any{"valid": false, "reason": "invalid_signature"}
	}
	entry := r.TrustRegistry.Resolve(presentation.DID)
	if entry == nil {
		return map[string]any{"valid": false, "reason": "unknown_did"}
	}
	return map[string]any{
		"valid":      true,
		"did":        presentation.DID,
		"agent_id":   presentation.AgentID,
		"scopes":     entry.Scopes,
		"trust_level": entry.TrustLevel,
	}
}

type KeyPair struct {
	PublicKey  string
	PrivateKey string
}

type AgentKeyStore struct {
	AlpDir string
	Keys   map[string]*KeyPair
}

func NewAgentKeyStore(alpDir string) *AgentKeyStore {
	ks := &AgentKeyStore{
		AlpDir: alpDir,
		Keys:   make(map[string]*KeyPair),
	}
	ks.Load()
	return ks
}

func (k *AgentKeyStore) identityDir() string {
	return filepath.Join(k.AlpDir, ".identity")
}

func (k *AgentKeyStore) keysPath() string {
	return filepath.Join(k.identityDir(), "agent_keys.json")
}

func (k *AgentKeyStore) Load() {
	p := k.keysPath()
	data, err := os.ReadFile(p)
	if err != nil {
		return
	}
	_ = data
}

func (k *AgentKeyStore) Save() {
	d := k.identityDir()
	if err := os.MkdirAll(d, 0755); err != nil {
		return
	}
	_ = d
}

func (k *AgentKeyStore) StoreKey(did, publicKey, privateKey string) {
	k.Keys[did] = &KeyPair{PublicKey: publicKey, PrivateKey: privateKey}
	k.Save()
}

func (k *AgentKeyStore) GetKey(did string) *KeyPair {
	return k.Keys[did]
}

func (k *AgentKeyStore) RemoveKey(did string) bool {
	if _, ok := k.Keys[did]; ok {
		delete(k.Keys, did)
		k.Save()
		return true
	}
	return false
}

func GenerateKeypair() *KeyPair {
	privateKey := strings.ReplaceAll(time.Now().Format(time.RFC3339Nano)+"-"+randomString(32), ":", "")
	publicKey := sha256Hex(privateKey)
	return &KeyPair{PublicKey: publicKey, PrivateKey: privateKey}
}

func CreateDID(agentID, publicKey string) string {
	keyHash := sha256Hex(publicKey)[:16]
	return "did:alp:" + agentID + ":" + keyHash
}

func sha256Hex(input string) string {
	h := sha256.Sum256([]byte(input))
	return hex.EncodeToString(h[:])
}

func randomString(n int) string {
	const letters = "abcdef0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[time.Now().UnixNano()%int64(len(letters))]
	}
	return string(b)
}

File: sdk/rust/src/identity.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

pub struct AgentIdentity {
    pub did: String,
    pub agent_id: String,
    pub public_key: String,
    pub created_at: String,
    pub metadata: HashMap<String, serde_json::Value>,
}

impl AgentIdentity {
    pub fn new(
        did: impl Into<String>,
        agent_id: impl Into<String>,
        public_key: impl Into<String>,
    ) -> Self {
        Self {
            did: did.into(),
            agent_id: agent_id.into(),
            public_key: public_key.into(),
            created_at: timestamp_now(),
            metadata: HashMap::new(),
        }
    }

    pub fn to_dict(&self) -> HashMap<String, serde_json::Value> {
        let mut dict = HashMap::new();
        dict.insert("did".into(), serde_json::Value::String(self.did.clone()));
        dict.insert(
            "agent_id".into(),
            serde_json::Value::String(self.agent_id.clone()),
        );
        dict.insert(
            "public_key".into(),
            serde_json::Value::String(self.public_key.clone()),
        );
        dict.insert(
            "created_at".into(),
            serde_json::Value::String(self.created_at.clone()),
        );
        dict.insert(
            "metadata".into(),
            serde_json::Value::Object(self.metadata.clone().collect()),
        );
        dict
    }
}

pub struct VerifiablePresentation {
    pub did: String,
    pub agent_id: String,
    pub claims: HashMap<String, serde_json::Value>,
    pub signature: String,
    pub issued_at: String,
}

impl VerifiablePresentation {
    pub fn new(
        did: impl Into<String>,
        agent_id: impl Into<String>,
        claims: HashMap<String, serde_json::Value>,
        signature: impl Into<String>,
    ) -> Self {
        Self {
            did: did.into(),
            agent_id: agent_id.into(),
            claims,
            signature: signature.into(),
            issued_at: timestamp_now(),
        }
    }

    pub fn verify(&self, public_key: &str) -> bool {
        let payload = serde_json::json!({
            "did": self.did,
            "agent_id": self.agent_id,
            "claims": self.claims,
        });
        let payload_str = payload.to_string();
        let expected = simple_hash(&(payload_str + public_key));
        self.signature == expected
    }

    pub fn sign(&mut self, private_key: &str) -> String {
        let payload = serde_json::json!({
            "did": self.did,
            "agent_id": self.agent_id,
            "claims": self.claims,
        });
        let payload_str = payload.to_string();
        self.signature = simple_hash(&(payload_str + private_key));
        self.signature.clone()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrustEntry {
    pub agent_id: String,
    pub scopes: Vec<String>,
    pub trust_level: String,
    pub registered_at: String,
}

impl TrustEntry {
    pub fn new(
        agent_id: impl Into<String>,
        scopes: Vec<String>,
        trust_level: impl Into<String>,
    ) -> Self {
        Self {
            agent_id: agent_id.into(),
            scopes,
            trust_level: trust_level.into(),
            registered_at: timestamp_now(),
        }
    }
}

pub struct TrustRegistry {
    alp_dir: String,
    entries: HashMap<String, TrustEntry>,
}

impl TrustRegistry {
    pub fn new(alp_dir: impl Into<String>) -> Self {
        let alp_dir = alp_dir.into();
        let mut registry = Self {
            alp_dir: alp_dir.clone(),
            entries: HashMap::new(),
        };
        registry.load();
        registry
    }

    fn identity_dir(&self) -> String {
        Path::new(&self.alp_dir)
            .join(".identity")
            .to_string_lossy()
            .into_owned()
    }

    fn trust_path(&self) -> String {
        Path::new(&self.identity_dir())
            .join("trust_registry.json")
            .to_string_lossy()
            .into_owned()
    }

    pub fn load(&mut self) {
        let path = self.trust_path();
        if !std::path::Path::new(&path).exists() {
            return;
        }
        if let Ok(content) = std::fs::read_to_string(&path) {
            if let Ok(entries) = serde_json::from_str::<HashMap<String, TrustEntry>>(&content) {
                self.entries = entries;
            }
        }
    }

    pub fn save(&self) {
        let dir = self.identity_dir();
        let _ = std::fs::create_dir_all(&dir);
        let path = self.trust_path();
        if let Ok(json) = serde_json::to_string_pretty(&self.entries) {
            let _ = std::fs::write(path, json);
        }
    }

    pub fn register(
        &mut self,
        did: impl Into<String>,
        agent_id: impl Into<String>,
        scopes: Vec<String>,
        trust_level: impl Into<String>,
    ) -> TrustEntry {
        let entry = TrustEntry::new(agent_id, scopes, trust_level);
        let did = did.into();
        self.entries.insert(did.clone(), entry.clone());
        self.save();
        entry
    }

    pub fn resolve(&self, did: &str) -> Option<&TrustEntry> {
        self.entries.get(did)
    }

    pub fn revoke(&mut self, did: &str) -> bool {
        if self.entries.remove(did).is_some() {
            self.save();
            true
        } else {
            false
        }
    }

    pub fn list_dids(&self) -> Vec<String> {
        self.entries.keys().cloned().collect()
    }

    pub fn has_scope(&self, did: &str, required_scope: &str) -> bool {
        if let Some(entry) = self.entries.get(did) {
            entry.scopes.contains(&required_scope.to_string())
        } else {
            false
        }
    }
}

pub struct IdentityResolver {
    trust_registry: std::rc::Rc<std::cell::RefCell<TrustRegistry>>,
}

impl IdentityResolver {
    pub fn new(trust_registry: std::rc::Rc<std::cell::RefCell<TrustRegistry>>) -> Self {
        Self { trust_registry }
    }

    pub fn verify_presentation(
        &self,
        presentation: &VerifiablePresentation,
        public_key: &str,
    ) -> HashMap<String, serde_json::Value> {
        let mut result = HashMap::new();
        if !presentation.verify(public_key) {
            result.insert("valid".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("invalid_signature".into()),
            );
            return result;
        }
        let registry = self.trust_registry.borrow();
        if let Some(entry) = registry.resolve(&presentation.did) {
            result.insert("valid".into(), serde_json::Value::Bool(true));
            result.insert(
                "did".into(),
                serde_json::Value::String(presentation.did.clone()),
            );
            result.insert(
                "agent_id".into(),
                serde_json::Value::String(presentation.agent_id.clone()),
            );
            result.insert(
                "scopes".into(),
                serde_json::Value::Array(
                    entry
                        .scopes
                        .iter()
                        .map(|s| serde_json::Value::String(s.clone()))
                        .collect(),
                ),
            );
            result.insert(
                "trust_level".into(),
                serde_json::Value::String(entry.trust_level.clone()),
            );
        } else {
            result.insert("valid".into(), serde_json::Value::Bool(false));
            result.insert(
                "reason".into(),
                serde_json::Value::String("unknown_did".into()),
            );
        }
        result
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyPair {
    pub public_key: String,
    pub private_key: String,
}

impl KeyPair {
    pub fn new(public_key: impl Into<String>, private_key: impl Into<String>) -> Self {
        Self {
            public_key: public_key.into(),
            private_key: private_key.into(),
        }
    }
}

pub struct AgentKeyStore {
    alp_dir: String,
    keys: HashMap<String, KeyPair>,
}

impl AgentKeyStore {
    pub fn new(alp_dir: impl Into<String>) -> Self {
        let alp_dir = alp_dir.into();
        let mut store = Self {
            alp_dir: alp_dir.clone(),
            keys: HashMap::new(),
        };
        store.load();
        store
    }

    fn identity_dir(&self) -> String {
        Path::new(&self.alp_dir)
            .join(".identity")
            .to_string_lossy()
            .into_owned()
    }

    fn keys_path(&self) -> String {
        Path::new(&self.identity_dir())
            .join("agent_keys.json")
            .to_string_lossy()
            .into_owned()
    }

    pub fn load(&mut self) {
        let path = self.keys_path();
        if !std::path::Path::new(&path).exists() {
            return;
        }
        if let Ok(content) = std::fs::read_to_string(&path) {
            if let Ok(keys) = serde_json::from_str::<HashMap<String, KeyPair>>(&content) {
                self.keys = keys;
            }
        }
    }

    pub fn save(&self) {
        let dir = self.identity_dir();
        let _ = std::fs::create_dir_all(&dir);
        let path = self.keys_path();
        if let Ok(json) = serde_json::to_string_pretty(&self.keys) {
            let _ = std::fs::write(path, json);
        }
    }

    pub fn store_key(
        &mut self,
        did: impl Into<String>,
        public_key: impl Into<String>,
        private_key: impl Into<String>,
    ) {
        let did = did.into();
        let pair = KeyPair::new(public_key, private_key);
        self.keys.insert(did.clone(), pair);
        self.save();
    }

    pub fn get_key(&self, did: &str) -> Option<&KeyPair> {
        self.keys.get(did)
    }

    pub fn remove_key(&mut self, did: &str) -> bool {
        if self.keys.remove(did).is_some() {
            self.save();
            true
        } else {
            false
        }
    }
}

pub fn generate_keypair() -> KeyPair {
    let ts = timestamp_now();
    let private_key = format!("{}-{}", ts, simple_hash(&ts));
    let public_key = simple_hash(&private_key);
    KeyPair::new(public_key, private_key)
}

pub fn create_did(agent_id: impl Into<String>, public_key: impl Into<String>) -> String {
    let public_key = public_key.into();
    let key_hash = simple_hash(&public_key);
    format!(
        "did:alp:{}:{}",
        agent_id.into(),
        &key_hash[..16.min(key_hash.len())]
    )
}

fn timestamp_now() -> String {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    now.as_millis().to_string()
}

fn simple_hash(input: &str) -> String {
    let mut hash: u64 = 0;
    for byte in input.bytes() {
        hash = hash.wrapping_mul(31).wrapping_add(byte as u64);
    }
    format!("{:x}", hash)
}
